### PR TITLE
feat: sort models list by "the last 5 used", then alphabetically

### LINKED ISSUE #143: Refactor `<select>` menus so we can have grouped options and fuzzy search
REF #124

### PR DESCRIPTION
**Pull request by GregoMac1**
Also closes #143 

image

### CODE REVIEW

**Comment by GregoMac1** [comment]
@fmaclen I forgot adding the `<Badge />` next to each model, you don't have to review this yet 😅

File: src/lib/components/FieldSelect.svelte
Comment (GregoMac1) [inline src/lib/components/FieldSelect.svelte:153]: @fmaclen maybe we could extract `<Combobox.Item>` to a separate component so we remove these duplicated lines, but I don't know if it's worth it.

File: src/lib/sessions.ts
Comment (fmaclen) [inline src/lib/sessions.ts:90]: We are only using this function in `FieldSelectModel` so let's keep this logic there, maybe rename it to `handleRecentlyUsedModels` so it's consistent with `handleChange()` and the label on the screen.

If we find we need to use this in other places later we might end up moving it back to `sessions.ts`.

File: src/lib/components/FieldSelectModel.svelte
Comment (fmaclen) [inline src/lib/components/FieldSelectModel.svelte:22]: - [ ] If there are no recent models it renders an empty list. If a group doesn't have any results we should include an empty message: `No recently used models`. I thought of suggesting hiding the group entirely but I feel like it's better UX if the user knows the feature even exists before they use a model for the first time.
  image

- [ ] The models in the recent list are duplicated in the other list, so when you type a model name you see the result twice. We probably shouldn't include the model in the 2nd group.
  image

- [ ] Alphabetical sorting got lost in the first group:
  image

File: src/lib/components/FieldSelect.svelte
Comment (fmaclen) [reply src/lib/components/FieldSelect.svelte:153]: Yeah, we could DRY the `<Combobox.Item>` but I think your current version looks more readable. Let's keep it 👍

File: src/i18n/en/index.ts
Comment (fmaclen) [inline src/i18n/en/index.ts:4]: Let's change this to `All other models` so the description matches the new behavior.

File: src/lib/components/FieldSelect.svelte
Comment (fmaclen) [inline src/lib/components/FieldSelect.svelte:198]: Here are some styles for ya:

```suggestion
	:global(.field-combobox-item-label) {
		@apply w-full grid grid-cols-[auto,max-content];
	}
```

image

File: src/lib/components/FieldSelect.svelte
Comment (fmaclen) [inline src/lib/components/FieldSelect.svelte:19]: Let's rename `extra` > `badge` so it's easier to _guess_ what component the value will be rendered in.

File: src/lib/components/FieldSelect.svelte
Comment (fmaclen) [inline src/lib/components/FieldSelect.svelte:78]: Why was `requestAnimationFrame` needed?

**Review by fmaclen — CHANGES_REQUESTED** [review]
Man, I've been looking forward to this feature!

PR looks good so far. Left you some minor issues and a CSS suggestion to fix the styling a bit. Needs some tests 🧪

File: src/i18n/es/index.ts
Comment (fmaclen) [inline src/i18n/es/index.ts:84]: Were these missing from my PR?

I think we need to add a CI check for this i18n stuff, it's easy to forget to translate stuff and/or re-generate the types.

File: src/lib/components/FieldSelect.svelte
Comment (GregoMac1) [reply src/lib/components/FieldSelect.svelte:198]: I made it look like this at the beginning because it looks good on the New session screen, but in the Settings screen I'm not sure if it looks nice because the input is wider. What do you think?

image

File: src/lib/components/FieldSelectModel.svelte
Comment (GregoMac1) [reply src/lib/components/FieldSelectModel.svelte:22]: > Alphabetical sorting got lost in the first group

Actually, I did this on purpose. I thought it'd be better to have the recently used models sorted by last used. Then, in the second group, models are sorted alphabetically. Should both groups be sorted alphabetically?

Comment (fmaclen) [reply src/lib/components/FieldSelectModel.svelte:22]: Ah, interesting. I'm not sure what the best sorting way should be, I'll have to to play with it.
Let's keep your current sorting by frequency, sounds like a good bet.

Make sure we add a couple of assertions for that sorting logic.

File: src/i18n/en/index.ts
Comment (GregoMac1) [reply src/i18n/en/index.ts:4]: Okay, so the recently used models shouldn't be included in the second group of models?

Comment (fmaclen) [reply src/i18n/en/index.ts:4]: Correct

**Comment by GregoMac1** [comment]
All done! I still have to add some tests.

File: src/lib/components/FieldSelectModel.svelte
Comment (fmaclen) [reply src/lib/components/FieldSelectModel.svelte]: Maybe we should reverse the order, in this screenshot you can see 2 of the 3 sessions with `codestral:latest` but `phi:latest` appears first:

image

Comment (fmaclen) [reply src/lib/components/FieldSelectModel.svelte]: I see, for some reason I always pictured this feature with the most popular at the top.
But that's okay, let's leave it with the current order.

Comment (GregoMac1) [reply src/lib/components/FieldSelectModel.svelte]: We could also sort the models by _most used_, but there'd be some details that we should consider:
- we should iterate over the sessions and count how many times was each model used. Then, we could sort the 5 most used models. However, if the user had a large number of sessions, we should see if this still performs well.
- let's suppose that my 5 most used models have ~100 sessions each. If I pulled some new models and preferred to use them from now on, they wouldn't appear in the first group of models until I have more sessions that use those newly pulled models.

Comment (fmaclen) [reply src/lib/components/FieldSelectModel.svelte]: You are right, there's quite a bit of complexity if we go the _most used_ route.
What we have now is already a huge UX improvement, we can punt these ideas for later.

File: src/lib/components/FieldSelect.svelte
Comment (fmaclen) [inline src/lib/components/FieldSelect.svelte:212]: Fixed the styling for mobile:

image

File: tests/settings.test.ts
Comment (fmaclen) [inline tests/settings.test.ts:239]: There is some duplication of tests with [`ui.test.ts`](https://github.com/fmaclen/hollama/blob/main/tests/ui.test.ts)
I think it makes more sense to move the non-duplicated tests to that file.

You can add a new `test.describe('FieldSelectModel', () => { ... });` section for the new tests.

Also, let's test these features in a the **New session** section, for example: `await page.goto('/sessions/qbhc0q');`, because we are likely to remove the models menu from the **Settings** page after #94

**Review by fmaclen — CHANGES_REQUESTED** [review]
Almost there, only minor code organization changes.

File: tests/ui.test.ts
Comment (GregoMac1) [inline tests/ui.test.ts:86]: I had to do this because If I did `await page.goto('/sessions/qbhc0q');` instead of going to `/settings` first, the `Run` button would never get enabled.

File: tests/ui.test.ts
Comment (fmaclen) [inline tests/ui.test.ts:127]: Let's add one minor assertion here to make sure the models are not duplicated across both groups.

```suggestion
		await expect(otherModelsGroup).toHaveCount(1);
		await expect(otherModelsGroup).toContainText(MOCK_API_TAGS_RESPONSE.models[2].name);
```

**Review by fmaclen — APPROVED** [review]
This looks great! 😎 
I left you a minor assertion suggestion but feel free to merge this after adding it.

File: tests/ui.test.ts
Comment (GregoMac1) [reply tests/ui.test.ts:86]: For some reason the `Run` button is still disabled when the tests are run in the GH Action. Debugging...

Comment (fmaclen) [reply tests/ui.test.ts:86]: Gotcha, yeah, let's get that sorted out before merging.

Comment (GregoMac1) [reply tests/ui.test.ts:86]: All tests pass if I set `retries: 2` in `playwright.config.ts` 🤔

**Pull request merged**
